crates/bevy_reflect/derive/src/enum_utility.rs

```diff
@@ -1,122 +1,263 @@
-            let field_ident = ident_or_index(field.data.ident.as_ref(), declare_index);
-            let field_ty = &field.data.ty;
-
-            let field_value = if field.attrs.ignore.is_ignored() {
-                match &field.attrs.default {
-                    DefaultBehavior::Func(path) => quote! { #path() },
-                    _ => quote! { #FQDefault::default() }
-                }
-            } else {
-                let field_accessor = match &field.data.ident {
-                    Some(ident) => {
-                        let name = ident.to_string();
-                        quote!(#ref_value.field(#name))
-                    }
-                    None => quote!(#ref_value.field_at(#reflect_index)),
-                };
-                reflect_index += 1;
-
-                let (resolve_error, resolve_missing) = if return_apply_error {
-                    let field_ref_str = match &field_ident {
-                        Member::Named(ident) => format!("{ident}"),
-                        Member::Unnamed(index) => format!(".{}", index.index)
-                    };
-                    let ty = field.data.ty.to_token_stream();
-
-                    (
-                        quote!(.ok_or(#bevy_reflect_path::ApplyError::MismatchedTypes {
-                            // The unwrap won't panic. By this point the #field_accessor would have been invoked once and any failure to
-                            // access the given field handled by the `resolve_missing` code bellow.
-                            from_type: ::core::convert::Into::into(
-                                #bevy_reflect_path::DynamicTypePath::reflect_type_path(#FQOption::unwrap(#field_accessor))
-                            ),
-                            to_type: ::core::convert::Into::into(<#ty as #bevy_reflect_path::TypePath>::type_path())
-                        })?),
-                        quote!(.ok_or(#bevy_reflect_path::ApplyError::MissingEnumField {
-                                variant_name: ::core::convert::Into::into(#name),
-                                field_name: ::core::convert::Into::into(#field_ref_str)
-                        })?)
-                    )
-                } else {
-                    (quote!(?), quote!(?))
+#[derive(Copy, Clone)]
+struct VariantField<'a, 'b> {
+    /// The pre-computed member for the field.
+    pub member: &'a Member,
+    /// The name of the variant that contains the field.
+    pub variant_name: &'a str,
+    /// The field data.
+    pub field: &'a StructField<'b>,
+}
+
+/// A builder for constructing [`EnumVariantOutputData`].
+///
+/// When a variant field needs to be constructed, the builder will:
+/// 1. Use `field_accessor` to access the field as an `Option<&dyn Reflect>`.
+/// 2. Use `field_unwrapper` to unwrap the field into a `&dyn Reflect`.
+/// 3. Use `field_constructor` to construct the field into a concrete value.
+struct EnumVariantOutputDataBuilder<'a> {
```

Comment:
I chose to use the builder pattern for this to also help improve readability (especially when you're not in a code editor and don't have access to hover annotations). Due to this, the default generators (i.e. `field_unwrapper` and `field_constructor`) will emit a compile error. 

In reality, this should never happen since the builder is only used within this module and we can ensure that all generators are given proper values.